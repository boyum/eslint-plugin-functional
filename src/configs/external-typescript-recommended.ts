import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import externalVanillaRecommended from "#eslint-plugin-functional/configs/external-vanilla-recommended";

const tsConfig = {
  "@typescript-eslint/prefer-readonly": "error",
  "@typescript-eslint/switch-exhaustiveness-check": "error",
} satisfies FlatConfig.Config["rules"];

export default {
  ...externalVanillaRecommended,
  ...tsConfig,
} satisfies FlatConfig.Config["rules"];
