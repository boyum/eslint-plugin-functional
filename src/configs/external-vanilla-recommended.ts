import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

export default {
  "prefer-const": "error",
  "no-param-reassign": "error",
  "no-var": "error",
} satisfies FlatConfig.Config["rules"];
