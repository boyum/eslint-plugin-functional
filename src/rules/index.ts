import * as functionalParameters from "./functional-parameters";
import * as immutableData from "./immutable-data";
import * as noClassInheritance from "./no-class-inheritance";
import * as noClasses from "./no-classes";
import * as noConditionalStatements from "./no-conditional-statements";
import * as noExpressionStatements from "./no-expression-statements";
import * as noLet from "./no-let";
import * as noLoopStatements from "./no-loop-statements";
import * as noMixedTypes from "./no-mixed-types";
import * as noPromiseReject from "./no-promise-reject";
import * as noReturnVoid from "./no-return-void";
import * as noThisExpressions from "./no-this-expressions";
import * as noThrowStatements from "./no-throw-statements";
import * as noTryStatements from "./no-try-statements";
import * as preferImmutableTypes from "./prefer-immutable-types";
import * as preferPropertySignatures from "./prefer-property-signatures";
import * as preferReadonlyTypes from "./prefer-readonly-type";
import * as preferTacit from "./prefer-tacit";
import * as readonlyType from "./readonly-type";
import * as typeDeclarationImmutability from "./type-declaration-immutability";

/**
 * All of the custom rules.
 */
export const rules: Readonly<{
  [functionalParameters.name]: typeof functionalParameters.rule;
  [immutableData.name]: typeof immutableData.rule;
  [noClasses.name]: typeof noClasses.rule;
  [noClassInheritance.name]: typeof noClassInheritance.rule;
  [noConditionalStatements.name]: typeof noConditionalStatements.rule;
  [noExpressionStatements.name]: typeof noExpressionStatements.rule;
  [noLet.name]: typeof noLet.rule;
  [noLoopStatements.name]: typeof noLoopStatements.rule;
  [noMixedTypes.name]: typeof noMixedTypes.rule;
  [noPromiseReject.name]: typeof noPromiseReject.rule;
  [noReturnVoid.name]: typeof noReturnVoid.rule;
  [noThisExpressions.name]: typeof noThisExpressions.rule;
  [noThrowStatements.name]: typeof noThrowStatements.rule;
  [noTryStatements.name]: typeof noTryStatements.rule;
  [preferImmutableTypes.name]: typeof preferImmutableTypes.rule;
  [preferPropertySignatures.name]: typeof preferPropertySignatures.rule;
  [preferReadonlyTypes.name]: typeof preferReadonlyTypes.rule;
  [preferTacit.name]: typeof preferTacit.rule;
  [readonlyType.name]: typeof readonlyType.rule;
  [typeDeclarationImmutability.name]: typeof typeDeclarationImmutability.rule;
}> = {
  [functionalParameters.name]: functionalParameters.rule,
  [immutableData.name]: immutableData.rule,
  [noClasses.name]: noClasses.rule,
  [noClassInheritance.name]: noClassInheritance.rule,
  [noConditionalStatements.name]: noConditionalStatements.rule,
  [noExpressionStatements.name]: noExpressionStatements.rule,
  [noLet.name]: noLet.rule,
  [noLoopStatements.name]: noLoopStatements.rule,
  [noMixedTypes.name]: noMixedTypes.rule,
  [noPromiseReject.name]: noPromiseReject.rule,
  [noReturnVoid.name]: noReturnVoid.rule,
  [noThisExpressions.name]: noThisExpressions.rule,
  [noThrowStatements.name]: noThrowStatements.rule,
  [noTryStatements.name]: noTryStatements.rule,
  [preferImmutableTypes.name]: preferImmutableTypes.rule,
  [preferPropertySignatures.name]: preferPropertySignatures.rule,
  [preferReadonlyTypes.name]: preferReadonlyTypes.rule,
  [preferTacit.name]: preferTacit.rule,
  [readonlyType.name]: readonlyType.rule,
  [typeDeclarationImmutability.name]: typeDeclarationImmutability.rule,
};
