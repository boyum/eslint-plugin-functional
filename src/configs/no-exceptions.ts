import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import { rules } from "#eslint-plugin-functional/rules";
import { ruleNameScope } from "#eslint-plugin-functional/utils/misc";

export default Object.fromEntries(
  Object.entries(rules)
    .filter(
      ([, rule]) =>
        rule.meta.deprecated !== true &&
        rule.meta.docs.category === "No Exceptions" &&
        rule.meta.docs.recommended !== false,
    )
    .map(([name, rule]) => [
      `${ruleNameScope}/${name}`,
      rule.meta.docs.recommendedSeverity,
    ]),
) satisfies FlatConfig.Config["rules"];
