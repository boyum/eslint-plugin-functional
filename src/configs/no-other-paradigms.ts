import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import { rules } from "#eslint-plugin-functional/rules";
import { ruleNameScope } from "#eslint-plugin-functional/utils/misc";

export default Object.fromEntries(
  Object.entries(rules)
    .filter(
      ([, rule]) =>
        rule.meta.deprecated !== true &&
        rule.meta.docs.category === "No Other Paradigms" &&
        rule.meta.docs.recommended !== false,
    )
    .map(([name, rule]) => [
      `${ruleNameScope}/${name}`,
      rule.meta.docs.recommendedSeverity,
    ]),
) satisfies FlatConfig.Config["rules"];
