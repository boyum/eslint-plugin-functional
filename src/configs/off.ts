import type { FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import { rules } from "#/rules";
import { ruleNameScope } from "#/utils/misc";

export default Object.fromEntries(
  Object.entries(rules).map(([name]) => [`${ruleNameScope}/${name}`, "off"]),
) satisfies FlatConfig.Config["rules"] as NonNullable<FlatConfig.Config["rules"]>;
