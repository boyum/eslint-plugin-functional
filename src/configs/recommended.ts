import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";
import { Immutability } from "is-immutable-type";

import { rules } from "#eslint-plugin-functional/rules";
import * as functionalParameters from "#eslint-plugin-functional/rules/functional-parameters";
import * as noConditionalStatements from "#eslint-plugin-functional/rules/no-conditional-statements";
import * as noLet from "#eslint-plugin-functional/rules/no-let";
import * as noThisExpressions from "#eslint-plugin-functional/rules/no-this-expressions";
import * as noThrowStatements from "#eslint-plugin-functional/rules/no-throw-statements";
import * as noTryStatements from "#eslint-plugin-functional/rules/no-try-statements";
import * as preferImmutableTypes from "#eslint-plugin-functional/rules/prefer-immutable-types";
import * as typeDeclarationImmutability from "#eslint-plugin-functional/rules/type-declaration-immutability";
import { RuleEnforcementComparator } from "#eslint-plugin-functional/rules/type-declaration-immutability";
import { ruleNameScope } from "#eslint-plugin-functional/utils/misc";

const recommended = Object.fromEntries(
  Object.entries(rules)
    .filter(
      ([, rule]) =>
        rule.meta.deprecated !== true &&
        rule.meta.docs.recommended === "recommended" &&
        rule.meta.docs.category !== "Stylistic",
    )
    .map(([name, rule]) => [
      `${ruleNameScope}/${name}`,
      rule.meta.docs.recommendedSeverity,
    ]),
) satisfies FlatConfig.Config["rules"];

const overrides = {
  [functionalParameters.fullName]: [
    "error",
    {
      enforceParameterCount: {
        ignoreLambdaExpression: true,
        ignoreIIFE: true,
        ignoreGettersAndSetters: true,
      },
    },
  ],
  [noConditionalStatements.fullName]: [
    "error",
    {
      allowReturningBranches: true,
    },
  ],
  [noLet.fullName]: [
    "error",
    {
      allowInForLoopInit: true,
    },
  ],
  [noThisExpressions.fullName]: "off",
  [noThrowStatements.fullName]: [
    "error",
    {
      allowInAsyncFunctions: true,
    },
  ],
  [noTryStatements.fullName]: "off",
  [preferImmutableTypes.fullName]: [
    "error",
    {
      enforcement: "None",
      ignoreInferredTypes: true,
      parameters: {
        enforcement: "ReadonlyDeep",
      },
    },
  ],
  [typeDeclarationImmutability.fullName]: [
    "error",
    {
      rules: [
        {
          identifiers: ["^I?Immutable.+"],
          immutability: Immutability.Immutable,
          comparator: RuleEnforcementComparator.AtLeast,
        },
        {
          identifiers: ["^I?ReadonlyDeep.+"],
          immutability: Immutability.ReadonlyDeep,
          comparator: RuleEnforcementComparator.AtLeast,
        },
        {
          identifiers: ["^I?Readonly.+"],
          immutability: Immutability.ReadonlyShallow,
          comparator: RuleEnforcementComparator.AtLeast,
          fixer: [
            {
              pattern: "^(Array|Map|Set)<(.+)>$",
              replace: "Readonly$1<$2>",
            },
            {
              pattern: "^(.+)$",
              replace: "Readonly<$1>",
            },
          ],
        },
        {
          identifiers: ["^I?Mutable.+"],
          immutability: Immutability.Mutable,
          comparator: RuleEnforcementComparator.AtMost,
          fixer: [
            {
              pattern: "^Readonly(Array|Map|Set)<(.+)>$",
              replace: "$1<$2>",
            },
            {
              pattern: "^Readonly<(.+)>$",
              replace: "$1",
            },
          ],
        },
      ],
    },
  ],
} satisfies FlatConfig.Config["rules"];

export default {
  ...recommended,
  ...overrides,
};
