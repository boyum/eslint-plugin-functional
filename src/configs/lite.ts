import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import * as functionalParameters from "#eslint-plugin-functional/rules/functional-parameters";
import * as immutableData from "#eslint-plugin-functional/rules/immutable-data";
import * as noConditionalStatements from "#eslint-plugin-functional/rules/no-conditional-statements";
import * as noExpressionStatements from "#eslint-plugin-functional/rules/no-expression-statements";
import * as preferImmutableTypes from "#eslint-plugin-functional/rules/prefer-immutable-types";

import recommended from "./recommended";

const overrides = {
  [functionalParameters.fullName]: [
    "error",
    {
      enforceParameterCount: false,
    },
  ],
  [immutableData.fullName]: ["error", { ignoreClasses: "fieldsOnly" }],
  [noConditionalStatements.fullName]: "off",
  [noExpressionStatements.fullName]: "off",
  [preferImmutableTypes.fullName]: [
    "error",
    {
      enforcement: "None",
      ignoreInferredTypes: true,
      parameters: {
        enforcement: "ReadonlyShallow",
      },
    },
  ],
} satisfies FlatConfig.Config["rules"];

export default {
  ...recommended,
  ...overrides,
};
