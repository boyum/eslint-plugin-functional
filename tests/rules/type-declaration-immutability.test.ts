import dedent from "dedent";
import { createRuleTester } from "eslint-vitest-rule-tester";
import { Immutability } from "is-immutable-type";
import { describe, expect, it } from "vitest";

import { name, rule } from "#/rules/type-declaration-immutability";

import { typescriptConfig } from "../utils/configs";

describe(name, () => {
  describe("typescript", () => {
    const options = [
      {
        rules: [
          {
            identifiers: ["^I?Immutable.+"],
            immutability: Immutability.Immutable,
            comparator: "AtLeast",
          },
          {
            identifiers: ["^I?ReadonlyDeep.+"],
            immutability: Immutability.ReadonlyDeep,
            comparator: "AtLeast",
          },
          {
            identifiers: ["^I?Readonly.+"],
            immutability: Immutability.ReadonlyShallow,
            comparator: "AtLeast",
          },
          {
            identifiers: ["^I?Mutable.+"],
            immutability: Immutability.Mutable,
            comparator: "AtMost",
          },
        ],
      },
    ];

    const { valid, invalid } = createRuleTester({
      name,
      rule,
      configs: typescriptConfig,
    });

    it("reports invalid shallow records", () => {
      const invalidResult1 = invalid({
        code: dedent`
          type ReadonlyFoo = {
            foo: number;
          };
        `,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult1.messages).toMatchSnapshot();

      const invalidResult2 = invalid({
        code: dedent`
          type ReadonlyFoo = {
            readonly foo: number;
            bar: {
              baz: string;
            };
          }
        `,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult2.messages).toMatchSnapshot();
    });

    it("reports invalid immutable/deep records", () => {
      const invalidResult1 = invalid({
        code: dedent`
          type ReadonlyDeepFoo = {
                      readonly foo: number;
                      readonly bar: {
                        baz: string;
                      };
                    };
        `,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult1.messages).toMatchSnapshot();

      const invalidResult2 = invalid({
        code: dedent`
          type ImmutableFoo = {
                      readonly foo: number;
                      readonly bar: {
                        baz: string;
                      };
                    };
        `,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult2.messages).toMatchSnapshot();
    });

    it("reports invalid shallow arrays", () => {
      const invalidResult = invalid({
        code: `type ReadonlyMyArray = Array<string>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it("reports invalid deep arrays", () => {
      const invalidResult = invalid({
        code: `type ReadonlyDeepMyArray = ReadonlyArray<{ foo: string; }>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it("reports invalid immutable arrays", () => {
      const invalidResult = invalid({
        code: `type ImmutableMyArray = ReadonlyArray<{ readonly foo: string; }>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it("reports invalid shallow sets", () => {
      const invalidResult = invalid({
        code: `type ReadonlyMySet = Set<string>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it("reports invalid deep sets", () => {
      const invalidResult = invalid({
        code: `type ReadonlyDeepMySet = ReadonlySet<{ foo: string; }>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it("reports invalid immutable sets", () => {
      const invalidResult = invalid({
        code: `type ImmutableMySet = ReadonlySet<{ readonly foo: string; }>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it("reports invalid shallow maps", () => {
      const invalidResult = invalid({
        code: `type ReadonlyMyMap = Map<string, string>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it("reports invalid deep maps", () => {
      const invalidResult = invalid({
        code: `type ReadonlyDeepMyMap = ReadonlyMap<string, { foo: string; }>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it("reports invalid immutable maps", () => {
      const invalidResult = invalid({
        code: `type ImmutableMyMap = ReadonlyMap<string, { readonly foo: string; }>;`,
        options,
        errors: ["AtLeast"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it("reports non-mutable primitives", () => {
      const invalidResult = invalid({
        code: "type MutableString = string;",
        options,
        errors: ["AtMost"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it("reports non-mutable records", () => {
      const invalidResult1 = invalid({
        code: dedent`
          type MutableFoo = {
            readonly foo: number;
          };
        `,
        options,
        errors: ["AtMost"],
      });
      expect(invalidResult1.messages).toMatchSnapshot();

      const invalidResult2 = invalid({
        code: dedent`
          type MutableFoo = Readonly<{
            foo: number;
          }>;
        `,
        options,
        errors: ["AtMost"],
      });
      expect(invalidResult2.messages).toMatchSnapshot();
    });

    it("doesn't report mutable sets", () => {
      valid({
        code: `type MutableSet = Set<string>;`,
        options,
      });
    });

    it("doesn't report mutable map", () => {
      valid({
        code: `type MutableMap = Map<string, string>;`,
        options,
      });
    });

    it("doesn't report valid types", () => {
      valid({
        code: "type ReadonlyString = string;",
        options,
      });
      valid({
        code: "type ReadonlyFoo = { readonly foo: number };",
        options,
      });
      valid({
        code: "type ReadonlyFoo = Readonly<{ foo: number }>;",
        options,
      });
      valid({
        code: "type ReadonlyFoo = { readonly foo: number; readonly bar: { baz: string; }; };",
        options,
      });
      valid({
        code: "type ReadonlySet = ReadonlySet<string>;",
        options,
      });
      valid({
        code: "type ReadonlyMap = ReadonlyMap<string, string>;",
        options,
      });
      valid({
        code: "type ReadonlyDeepString = string;",
        options,
      });
      valid({
        code: "type ReadonlyDeepFoo = { readonly foo: number; readonly bar: { readonly baz: string; }; };",
        options,
      });
      valid({
        code: "type ReadonlyDeepSet = ReadonlySet<string>;",
        options,
      });
      valid({
        code: "type ReadonlyDeepMap = ReadonlyMap<string, string>;",
        options,
      });
      valid({
        code: "type ReadonlyDeepSet = ReadonlySet<{ readonly foo: string; }>;",
        options,
      });
      valid({
        code: "type ReadonlyDeepMap = ReadonlyMap<string, { readonly foo: string; }>;",
        options,
      });
      valid({
        code: "type ImmutableString = string;",
        options,
      });
      valid({
        code: "type ImmutableFoo = { readonly foo: number; readonly bar: { readonly baz: string; }; };",
        options,
      });
      valid({
        code: "type ImmutableSet = Readonly<ReadonlySet<{ readonly foo: string; }>>;",
        options,
      });
      valid({
        code: "type ImmutableMap = Readonly<ReadonlyMap<string, { readonly foo: string; }>>;",
        options,
      });
      valid({
        code: "type MutableFoo = { foo: number };",
        options,
      });
      valid({
        code: "type MutableFoo = { readonly foo: number; bar: { readonly baz: string; }; };",
        options,
      });
      valid({
        code: "type MutableSet = Set<{ readonly foo: string; }>;",
        options,
      });
      valid({
        code: "type MutableMap = Map<string, { readonly foo: string; }>;",
        options,
      });
    });

    describe("ignoreInterfaces", () => {
      it("doesn't report interfaces when enabled", () => {
        valid({
          code: dedent`
            interface ReadonlyFoo {
              foo: number;
            }
          `,
          options: [{ ignoreInterfaces: true }],
        });
      });
    });

    describe("ignoreIdentifierPattern", () => {
      it("doesn't report interfaces when enabled", () => {
        valid({
          code: dedent`
            type ReadonlyFoo = {
              foo: number;
            };
          `,
          options: [
            {
              ignoreIdentifierPattern: "Foo",
            },
          ],
        });
      });
    });

    it("respects override settings", () => {
      valid({
        code: dedent`
          type ReadonlyDeepFoo = ReadonlyDeep<{ foo: { bar: string; }; }>;
          type ReadonlyDeep<T> = T;
        `,
        options,
        settings: {
          immutability: {
            overrides: [
              {
                type: "ReadonlyDeep",
                to: Immutability.ReadonlyDeep,
              },
            ],
          },
        },
      });

      const invalidResult = invalid({
        code: `type MutableSet = Set<string>;`,
        options,
        settings: {
          immutability: {
            overrides: {
              keepDefault: false,
              values: [
                {
                  type: { from: "lib", name: "Set" },
                  to: Immutability.Immutable,
                },
              ],
            },
          },
        },
        errors: ["AtMost"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });
  });
});
