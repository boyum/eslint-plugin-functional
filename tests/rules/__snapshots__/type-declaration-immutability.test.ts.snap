// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`type-declaration-immutability > typescript > reports invalid deep arrays 1`] = `
[
  {
    "column": 6,
    "endColumn": 25,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at least "ReadonlyDeep" (actual: "ReadonlyShallow").",
    "messageId": "AtLeast",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > reports invalid deep maps 1`] = `
[
  {
    "column": 6,
    "endColumn": 23,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at least "ReadonlyDeep" (actual: "ReadonlyShallow").",
    "messageId": "AtLeast",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > reports invalid deep sets 1`] = `
[
  {
    "column": 6,
    "endColumn": 23,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at least "ReadonlyDeep" (actual: "ReadonlyShallow").",
    "messageId": "AtLeast",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > reports invalid immutable arrays 1`] = `
[
  {
    "column": 6,
    "endColumn": 22,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
    "messageId": "AtLeast",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > reports invalid immutable maps 1`] = `
[
  {
    "column": 6,
    "endColumn": 20,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
    "messageId": "AtLeast",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > reports invalid immutable sets 1`] = `
[
  {
    "column": 6,
    "endColumn": 20,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at least "Immutable" (actual: "ReadonlyDeep").",
    "messageId": "AtLeast",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > reports invalid immutable/deep records 1`] = `
[
  {
    "column": 6,
    "endColumn": 21,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at least "ReadonlyDeep" (actual: "ReadonlyShallow").",
    "messageId": "AtLeast",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > reports invalid immutable/deep records 2`] = `
[
  {
    "column": 6,
    "endColumn": 18,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at least "Immutable" (actual: "ReadonlyShallow").",
    "messageId": "AtLeast",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > reports invalid shallow arrays 1`] = `
[
  {
    "column": 6,
    "endColumn": 21,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
    "messageId": "AtLeast",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > reports invalid shallow maps 1`] = `
[
  {
    "column": 6,
    "endColumn": 19,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
    "messageId": "AtLeast",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > reports invalid shallow records 1`] = `
[
  {
    "column": 6,
    "endColumn": 17,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
    "messageId": "AtLeast",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > reports invalid shallow records 2`] = `
[
  {
    "column": 6,
    "endColumn": 17,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
    "messageId": "AtLeast",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > reports invalid shallow sets 1`] = `
[
  {
    "column": 6,
    "endColumn": 19,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at least "ReadonlyShallow" (actual: "Mutable").",
    "messageId": "AtLeast",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > reports non-mutable primitives 1`] = `
[
  {
    "column": 6,
    "endColumn": 19,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at most "Mutable" (actual: "Immutable").",
    "messageId": "AtMost",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > reports non-mutable records 1`] = `
[
  {
    "column": 6,
    "endColumn": 16,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at most "Mutable" (actual: "Immutable").",
    "messageId": "AtMost",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > reports non-mutable records 2`] = `
[
  {
    "column": 6,
    "endColumn": 16,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at most "Mutable" (actual: "Immutable").",
    "messageId": "AtMost",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;

exports[`type-declaration-immutability > typescript > respects override settings 1`] = `
[
  {
    "column": 6,
    "endColumn": 16,
    "endLine": 1,
    "line": 1,
    "message": "This type is declare to have an immutability of at most "Mutable" (actual: "Immutable").",
    "messageId": "AtMost",
    "nodeType": "Identifier",
    "ruleId": "type-declaration-immutability",
    "severity": 2,
  },
]
`;
